pkg/forwarder/internal/retry/queue_duration_capacity.go

```diff
@@ -0,0 +1,159 @@
+type QueueDurationCapacity struct {
+	accumulators           map[string]*timeIntervalAccumulator
+	optionalDiskSpace      diskSpace
+	maxMemSizeInBytes      int
+	historyDuration        time.Duration
+	bucketDuration         time.Duration
+	queueDiskSpaceUsedList []QueueDiskSpaceUsed
+}
+
+type diskSpace interface {
+	computeAvailableSpace(extraSize int64) (int64, error)
+}
+
+// QueueDiskSpaceUsed provides a method to get the disk space used by the retry queue.
+type QueueDiskSpaceUsed interface {
+	GetDiskSpaceUsed() int64
+}
+
+// NewQueueDurationCapacity creates a new instance of *QueueDurationCapacity.
+// if `optionalDiskSpace` is not nil, the capacity also use the storage on disk.
+// `historyDuration` is the duration used to compute the number of bytes received per second.
+// `bucketDuration` is the size of a bucket.
+func NewQueueDurationCapacity(
+	historyDuration time.Duration,
+	bucketDuration time.Duration,
+	maxMemSizeInBytes int,
+	optionalDiskSpace diskSpace,
+	queueDiskSpaceUsedList []QueueDiskSpaceUsed) *QueueDurationCapacity {
+	if optionalDiskSpace != nil && reflect.ValueOf(optionalDiskSpace).IsNil() {
+		optionalDiskSpace = nil
+	}
+
+	return &QueueDurationCapacity{
+		accumulators:      make(map[string]*timeIntervalAccumulator),
+		maxMemSizeInBytes: maxMemSizeInBytes,
+		optionalDiskSpace: optionalDiskSpace,
+		historyDuration:   historyDuration,
+		bucketDuration:    bucketDuration,
+	}
+}
+
+// OnTransaction must be called for each transaction.
+// Note: because of alternateDomains, `mainDomain` is not necessary the same as transaction.Domain.
+func (r *QueueDurationCapacity) OnTransaction(
+	transaction *transaction.HTTPTransaction,
+	mainDomain string,
+	now time.Time) error {
+
+	var accumulator *timeIntervalAccumulator
+	var found bool
+	if accumulator, found = r.accumulators[mainDomain]; !found {
+		var err error
+		accumulator, err = newTimeIntervalAccumulator(r.historyDuration, r.bucketDuration)
+		if err != nil {
+			return err
+		}
+		r.accumulators[mainDomain] = accumulator
+	}
+
+	accumulator.add(now, int64(transaction.GetPayloadSize()))
+	return nil
+}
+
+// QueueCapacityStats represents statistics about the capacity of the retry queue.
+type QueueCapacityStats struct {
+	Capacity       time.Duration
+	BytesPerSec    float64
+	AvailableSpace int64
+}
+
+// ComputeCapacity computes the capacity of the retry queue express as a duration.
+// Return statistics by domain name.
+func (r *QueueDurationCapacity) ComputeCapacity(t time.Time) (map[string]QueueCapacityStats, error) {
+	diskSpace, err := r.getTotalDiskSpace()
+	if err != nil {
+		return nil, err
+	}
+
+	var totalBytesPerSec float64
+	for _, accumulator := range r.accumulators {
+		totalBytesPerSec += getSpeedRate(accumulator, t)
+	}
+
+	durations := make(map[string]QueueCapacityStats)
+	for domain, accumulator := range r.accumulators {
+		bytesPerSec := getSpeedRate(accumulator, t)
```

Comment:
I suspect that this could be easily made to call `getSpeedRate` once for each accumulator, saving a little CPU?